### PR TITLE
Feat: Implement Predictive Coding local Hebbian learning rules (#1555)

### LINKED ISSUE #1555: Implement Predictive Coding local Hebbian learning rules
## Summary

Implement the "slow" learning phase of Predictive Coding — local Hebbian-like weight update rules that adjust synaptic weights based on prediction errors after inference has settled. This is the key differentiator from standard backpropagation: updates depend only on locally available information (pre-synaptic activity and post-synaptic error).

Part of #1549
Depends on: #1554 (PC inference engine)

### 1. Create `src/predictiveCoding/PredictiveCodingLearning.ts`

Implement the PC weight update rule:

```
After inference has settled for a data sample:
  For each synapse with weight W connecting neuron j → neuron i:
    ΔW = η_learn · f'(a(i)) · ε(i) · x(j)
    
  Where:
    η_learn = PC learning rate (from config)
    f'(a(i)) = derivative of squash function at pre-activation of target neuron
    ε(i) = prediction error at target neuron (from inference)
    x(j) = activation of source neuron (latent value after settling)
```

This is a local Hebbian rule: each synapse update depends only on:
- The error at its target neuron
- The activation of its source neuron
- The local derivative of the activation function

### 2. Bias Updates

Similarly, bias updates use local prediction errors:
```
Δb(i) = η_learn · f'(a(i)) · ε(i)
```

### 3. Batch Accumulation

- Accumulate weight/bias gradients across a mini-batch of samples
- Apply averaged updates after the batch completes
- This maintains the two-timescale separation: many inference steps per weight update

### 4. Integration with Existing Weight Constraints

- Respect existing weight precision constraints (minimum 1 plank unit = 1e-7)
- Respect synapse type constraints (positive, negative, condition)
- Apply `maximumWeightAdjustmentScale` from existing training config if configured
- Ensure weight updates don't violate topology constraints

## Acceptance Criteria

- [ ] `PredictiveCodingLearning.ts` implements local Hebbian weight updates
- [ ] Bias updates implemented alongside weight updates
- [ ] Batch accumulation mode works correctly
- [ ] Weight constraints (precision, type, scale) respected
- [ ] Unit tests including:
  - Known gradient: hand-computed weight update matches implementation
  - Learning reduces energy: verify total prediction error decreases over training epochs
  - Weight constraints: verify updates respect precision and type bounds
  - Batch vs single-sample: verify batch averaging produces correct results
  - Squash derivative: verify correct derivatives for multiple activation functions (RELU, TANH, SIGMOID)
- [ ] All existing tests pass unchanged
- [ ] `./quality.sh` passes
- [ ] Australian English spelling throughout

## Coding Practices Reminder
- TDD: Write a test with hand-computed expected weight deltas before implementing the update rule
- DRY: Reuse existing squash function derivative implementations from `src/methods/activations/`
- KISS: Single learning rate for all layers initially; per-layer rates can be a future enhancement

## References
- Whittington & Bogacz (2017): "An Approximation of the Error Backpropagation Algorithm in a Predictive Coding Network with Local Hebbian Synaptic Plasticity"
- Millidge et al. (2022): "Predictive Coding Approximates Backprop Along Arbitrary Computation Graphs"

### PR DESCRIPTION
**Pull request by stservice**
## Summary

Implement the "slow" learning phase of Predictive Coding — local Hebbian-like
weight update rules that adjust synaptic weights based on prediction errors
after inference has settled. This is the key differentiator from standard
backpropagation: each synapse update depends only on locally available
information (pre-synaptic activity, post-synaptic error, and the local
activation derivative). Closes #1555.

### Weight update rule

```
ΔW(j→i) = η_learn · f'(a(i)) · ε(i) · x(j)
Δb(i)   = η_learn · f'(a(i)) · ε(i)
```

### Key features

- `computeWeightGradients()`: computes local Hebbian gradients from settled
  inference state, using the correct activation function derivative for each
  neuron
- `applyHebbianUpdate()`: applies gradients while respecting plank precision
  (1e-7), synapse type constraints (positive/negative/condition), and weight
  bounds
- Batch accumulation supported via gradient averaging across multiple samples
  before applying a single update

## Evidence

This is a backend/algorithmic change with no visual output. Correctness is
verified by 12 unit tests with hand-computed expected values and constraint
validation. All 4221 tests pass including the 12 new tests. `./quality.sh`
passes cleanly.

## Test Plan

New test file: `test/predictiveCoding/PredictiveCodingLearning.ts` (12 tests)

- **Hand-computed weight deltas (IDENTITY)**: verifies ΔW = η · 1 · ε · x
  against manually computed values
- **Hand-computed weight deltas (LOGISTIC)**: verifies correct σ(x)·(1-σ(x))
  derivative usage
- **Energy decreases over epochs**: verifies prediction error energy reduces
  with repeated learning
- **Plank precision**: weights below 1e-7 magnitude snap to zero
- **Positive synapse constraint**: positive-typed synapses remain ≥ 0 after
  aggressive updates
- **Negative synapse constraint**: negative-typed synapses remain ≤ 0 after
  aggressive updates
- **Condition synapse immutability**: condition-typed synapses are never
  modified
- **Batch averaging**: accumulated gradients produce correct mean deltas
- **Batch vs sequential**: batch mode produces different results from online
  mode (verifying distinct code paths)
- **TANH derivative**: verifies 1 - tanh²(a) derivative is used correctly
- **ReLU derivative**: verifies piecewise derivative (1 for positive, 0 for
  negative) is used correctly
- **Zero errors → zero gradients**: converged inference produces near-zero
  weight updates

---

## Automated Fix Details
This PR addresses issue #1555: **Implement Predictive Coding local Hebbian learning rules**

## Quality Checks
- Followed TDD methodology
- All new functionality has corresponding tests
- Ran `./quality.sh` - all checks pass

## Notes
- No existing tests were removed or commented out
- If any test modifications were required due to business logic changes, they are documented in the commits

Closes #1555

### CODE REVIEW

**Comment by stservice** [comment]
This is a false positive from the automated dependency review action. Both `js-sys` and `wasm-bindgen` are well-known Rust/WebAssembly ecosystem crates maintained by the [rustwasm](https://github.com/rustwasm) working group and are dual-licensed under **MIT / Apache-2.0**.

The dependency review action reports "Null" for their licenses because it cannot resolve license metadata from Cargo.toml version range specifiers — this is a known limitation of the GitHub dependency review action with Cargo/crate dependencies.

You can verify the licenses directly:
- [`wasm-bindgen` on crates.io](https://crates.io/crates/wasm-bindgen) — MIT/Apache-2.0
- [`js-sys` on crates.io](https://crates.io/crates/js-sys) — MIT/Apache-2.0

No code changes needed.

**Review by nleck — APPROVED** [review]
